s2/gen.go

```diff
@@ -0,0 +1,1590 @@
+// retval will contain emitted bytes, but can be nil if this is not interesting.
+// dstBase and litBase are updated.
+// Uses 2 GP registers. With AVX 4 registers.
+// If updateDst is true dstBase will have the updated end pointer and an additional register will be used.
+func emitLiteral(name string, litLen, retval, dstBase, litBase reg.GPVirtual, end LabelRef, avx, updateDst bool) {
+	n := GP32()
+	n16 := GP32()
+
+	// We always add litLen bytes
+	if retval != nil {
+		MOVL(litLen.As32(), retval.As32())
+	}
+	MOVL(litLen.As32(), n)
+
+	SUBL(U8(1), n.As32())
+	// Return if AX was 0
+	JC(end)
+
+	// Find number of bytes to emit for tag.
+	CMPL(n.As32(), U8(60))
+	JLT(LabelRef("one_byte_" + name))
+	CMPL(n.As32(), U32(1<<8))
+	JLT(LabelRef("two_bytes_" + name))
+	CMPL(n.As32(), U32(1<<16))
+	JLT(LabelRef("three_bytes_" + name))
+	CMPL(n.As32(), U32(1<<24))
+	JLT(LabelRef("four_bytes_" + name))
+
+	Label("five_bytes_" + name)
+	MOVB(U8(252), Mem{Base: dstBase})
+	MOVL(n.As32(), Mem{Base: dstBase, Disp: 1})
+	if retval != nil {
+		ADDQ(U8(5), retval)
+	}
+	ADDQ(U8(5), dstBase)
+	JMP(LabelRef("memmove_" + name))
+
+	Label("four_bytes_" + name)
+	MOVL(n, n16)
+	SHRL(U8(16), n16.As32())
+	MOVB(U8(248), Mem{Base: dstBase})
+	MOVW(n.As16(), Mem{Base: dstBase, Disp: 1})
+	MOVB(n16.As8(), Mem{Base: dstBase, Disp: 3})
+	if retval != nil {
+		ADDQ(U8(4), retval)
+	}
+	ADDQ(U8(4), dstBase)
+	JMP(LabelRef("memmove_" + name))
+
+	Label("three_bytes_" + name)
+	MOVB(U8(0xf4), Mem{Base: dstBase})
+	MOVW(n.As16(), Mem{Base: dstBase, Disp: 1})
+	if retval != nil {
+		ADDQ(U8(3), retval)
+	}
+	ADDQ(U8(3), dstBase)
+	JMP(LabelRef("memmove_" + name))
+
+	Label("two_bytes_" + name)
+	MOVB(U8(0xf0), Mem{Base: dstBase})
+	MOVB(n.As8(), Mem{Base: dstBase, Disp: 1})
+	if retval != nil {
+		ADDQ(U8(2), retval)
+	}
+	ADDQ(U8(2), dstBase)
+	JMP(LabelRef("memmove_" + name))
+
+	Label("one_byte_" + name)
+	SHLB(U8(2), n.As8())
+	MOVB(n.As8(), Mem{Base: dstBase})
+	if retval != nil {
+		ADDQ(U8(1), retval)
+	}
+	ADDQ(U8(1), dstBase)
+	// Fallthrough
+
+	Label("memmove_" + name)
+
+	// copy(dst[i:], lit)
+	if true {
+		dstEnd := GP64()
+		if updateDst {
+			LEAQ(Mem{Base: dstBase, Index: litLen, Scale: 1}, dstEnd)
+		}
+		length := GP64()
+		MOVL(litLen.As32(), length.As32())
+		genMemMove2("emit_lit_memmove_"+name, dstBase, litBase, length, end, avx)
+		if updateDst {
+			MOVQ(dstEnd, dstBase)
+		}
+	} else {
+		genMemMove("emit_lit_memmove_"+name, dstBase, litBase, litLen, end)
+	}
+	return
+}
+
+// genEmitRepeat generates a standlone emitRepeat.
+func genEmitRepeat() {
+	TEXT("emitRepeat", NOSPLIT, "func(dst []byte, offset, length int) int")
+	Doc("emitRepeat writes a repeat chunk and returns the number of bytes written.",
+		"Length must be at least 4 and < 1<<32", "")
+	Pragma("noescape")
+
+	dstBase, offset, length, retval := GP64(), GP64(), GP64(), GP64()
+
+	// retval = 0
+	XORQ(retval, retval)
+
+	Load(Param("dst").Base(), dstBase)
```

Comment:
Perhaps do:

```go
dstBase := Load(Param("dst").Base(), GP64())
```

and remove the `dstBase` assignment earlier?

Reply:
Load returns a `reg.Register` and I want to send a `reg.GPVirtual`, so not possible.